chia/cmds/db_upgrade_func.py

```diff
@@ -0,0 +1,301 @@
+# configuration file will not be updated to use the new database. Only when using
+# the currently configured db file, and writing to the default output file will
+# the configuration file also be updated
+def db_upgrade_func(
+    root_path: Path,
+    in_db_path: Optional[Path] = None,
+    out_db_path: Optional[Path] = None,
+    no_update_config: bool = False,
+):
+
+    update_config: bool = in_db_path is None and out_db_path is None and not no_update_config
+
+    config: Dict
+    selected_network: str
+    db_pattern: str
+    if in_db_path is None or out_db_path is None:
+        config = load_config(root_path, "config.yaml")["full_node"]
+        selected_network = config["selected_network"]
+        db_pattern = config["database_path"]
+
+    db_path_replaced: str
+    if in_db_path is None:
+        db_path_replaced = db_pattern.replace("CHALLENGE", selected_network)
+        in_db_path = path_from_root(root_path, db_path_replaced)
+
+    if out_db_path is None:
+        db_path_replaced = db_pattern.replace("CHALLENGE", selected_network).replace("_v1_", "_v2_")
+        out_db_path = path_from_root(root_path, db_path_replaced)
+        mkdir(out_db_path.parent)
+
+    asyncio.run(convert_v1_to_v2(in_db_path, out_db_path))
+
+    if update_config:
+        print("updating config.yaml")
+        config = load_config(root_path, "config.yaml")
+        new_db_path = db_pattern.replace("_v1_", "_v2_")
+        config["full_node"]["database_path"] = new_db_path
+        print(f"database_path: {new_db_path}")
+        save_config(root_path, "config.yaml", config)
+
+    print(f"\n\nLEAVING PREVIOUS DB FILE UNTOUCHED {in_db_path}\n")
+
+
+BLOCK_COMMIT_RATE = 5000
+SES_COMMIT_RATE = 1000
+HINT_COMMIT_RATE = 1000
+COIN_COMMIT_RATE = 15000
+
+
+async def convert_v1_to_v2(in_path: Path, out_path: Path) -> None:
+    import aiosqlite
+    from chia.util.db_wrapper import DBWrapper
+
+    if out_path.exists():
+        print(f"output file already exists. {out_path}")
+        raise RuntimeError("already exists")
+
+    print(f"opening file for reading: {in_path}")
+    async with aiosqlite.connect(in_path) as in_db:
+        try:
+            async with in_db.execute("SELECT * from database_version") as cursor:
+                row = await cursor.fetchone()
+                if row is not None and row[0] != 1:
+                    print(f"blockchain database already version {row[0]}\nDone")
+                    raise RuntimeError("already v2")
+        except aiosqlite.OperationalError:
+            pass
+
+        store_v1 = await BlockStore.create(DBWrapper(in_db, db_version=1))
+
+        print(f"opening file for writing: {out_path}")
+        async with aiosqlite.connect(out_path) as out_db:
+            await out_db.execute("pragma journal_mode=OFF")
+            await out_db.execute("pragma synchronous=OFF")
+            await out_db.execute("pragma cache_size=1000000")
```

Comment:
This will allow up to 1,000,000 * page_size to be allocated in memory [[info](https://www.sqlite.org/pragma.html#pragma_cache_size)], which would be ~3.8 GiB.
My RPi with 8G Memory survives the upgrade run with this configuration, but I have also a very big swapfile active.
Not sure what the min. HW requirements for chia are.

Is it possible to make this dynamic according to the available systems memory?

Reply:
it's probably a good idea to dial that down a bit. I'll do that in a separate PR

Reply:
Actually. Is it really necessary to keep data cached in the out_db?
On low end RPi I had best results when completely removing this parameter.

Reply:
btw, same question regarding `pragma temp_store` .. I don't see the need for out_db, but it increases memory usage. I'm not 100% certain why, because there is no temp table and I droped all indexe, could be from primary keys.

Reply:
Apparently I was wrong with my statement about cache_size. 
I read in some blogs that it would make sense to set the cache_size as big as the transaction chunk will be, to minimize memory fragmentation . I've already tested that on my plotter HW (128 GB memory) and it helped in throughput.

The challenge will be to use this approach also for low end HW.
Here is a chart from my first tests a few days ago, which lead me to the conclusion to not use cache_size (because it looks like it doesn't release memory after a commit).
![grafik](https://user-images.githubusercontent.com/83925572/150173486-f403270e-90f8-4622-a1e6-a511a0b6f184.png)

I'm currently testing `PRAGMA shrink_memory` as a possible solution.

Reply:
what about just setting it to something smaller, like 500 MiB?

Reply:
shrink_memory is promising.

left try (1.) is again using 2 GB cache_size not releasing memory, which I aborted after some time. 
right side (2.) is executing `pragma shrink_memory` after every commit.
![grafik](https://user-images.githubusercontent.com/83925572/150179012-ea266ee7-2dea-4a3d-a3ed-3e0756d632ab.png)

you see increase in swap usage, this is where the sess commit rate needs to be reduced in the first place. and than I forgot to add shrink_memory at one place. I'll definitely retry that again with these adjustments.

coin_store convertion is currently running with ~11600 coins/s after 18M rows. 
On my tries the [last time](https://github.com/Chia-Network/chia-blockchain/pull/9613#issuecomment-1014921886), the best was 6300 coins/s after 38M.

Reply:
500 MiB would work as well. For systems like RPi with 4GB this will probably be a proper way.
How I understand it, then the commit rates should be reduced as well so one transaction can fit into the cache. 
(I'm using the commit rates I worte up in my summary below).

Reply:
Finally I figured out from where all the reserved memory comes from when starting the upgrade. 
It looks like there are some big LRUCaches implicitly allocated, when calling CoinStore and BlockStore for the table intialization. For testing purpose I commented that out and now I am able to actually use the cache_size as meant.

To improve migration performance, I'm still doing some tests on HW < 8GB memory figuring out a good cache/commit ratio. And there is a sqlite forum feedback, about vacuum as migration preparation from [Mr. sqlite himself](https://sqlite.org/forum/forumpost/61ced9482a) which I want to try out as well.

How is the migration timeline? Is it actually worth my effort? as I saw this is already merged into main.

Reply:
> How is the migration timeline?

My understanding (and hope) is that we will have nailed down the v2 schema for the upcoming 1.3 release, but still default to v1. v2 is still considered experimental. But to encourage testing of it, once we feel confident we switch over to default new installs to v2 and encourage people to convert. Hopefully there v2 merits will do the bulk of that encouragement.

>  Is it actually worth my effort?

Yes, once we release 1.3, I would like the migration experience to be as smooth as possible for people wanting to try it out.

>  as I saw this is already merged into main.

and I'm still hoping to make changes to the v2 schema. See: https://github.com/Chia-Network/chia-blockchain/pull/9778

Reply:
@arvidn 
I rewrote the migration script and doing final tests right now. 
Tried to focus on low-end HW and unvacuumed databases, which I think are operated by most farmers. I guess you can maybe use parts of it for cherry-picking, because due to my lack of programming exp it may look a bit weird.
When I've finished my tests, what would you prefer? 
Should I just send it over to you directly? Or create a PR on main or a branch?

Reply:
I think a PR would be best. I can branch off and make changes to that easily

Reply:
@arvidn 
btw, you should really consider removing the primary key like suggested by @xorinox for the tables and replace it by unique index. 
Initially I didn't like this proposal because, coming from Oracle databases, I'd consider such changes as bad practice.
But in the meantime I read in several blog posts and in the [documentation](https://www.sqlite.org/lang_createtable.html) as well, sqlite handles the Primary Key for an ordinary table just like if you create an Unique Index.

To give you an impression, my current upgrade tests on RPi4-8GB, DB@SSD, to peak 1361584
Runtime with Primary Key: ~140m
Runtime with Unique Index instead: ~80m

I'll probably be able to create the PR within the next few days, so you can check the other changes I made.

Reply:
I find it a bit scary to use writable schemas, and (possibly) duplicating large tables. But I'm also not particularly experienced with sqlite.

Reply:
I don't have much exp in sqlite myself and didn't want to suggest using writable_schemas. I'd consider using that option only when advised by DB vendor or when already in an emergency situation. 

The inflexibility of PKs (no disable/enable/integrity check/no ddl change) in sqlite just let me think, why not replace it with unique keys, if there is no actual difference and gain that flexibility during migrations. Even though I still don't like it to suggest such a clunky workaround removing PKs.

So I created a new [sqlite forum thread](https://sqlite.org/forum/forumpost/3959ee1f45) to ask the more expierenced sqlite users if I oversee something. Some user didn't understand my intention and some answered (imo) incorrectly, but the one I linked made a very good point.

Instead of removing PK, use a [WITHOUT ROWID](https://www.sqlite.org/withoutrowid.html) table definition for tables where it fits and for bulk insert use `order by <primary_key_col>` to get best performance.

Using the second suggestion(order by) already gives some bulk load performance improvements for ordinary tables. The first, using WITHOUT ROWID, need some additional testing, at least on my side. 
I don't know if you have already considered/tested WITHOUT ROWID on tables?

Reply:
I ran some benchmarks using `WITHOUT ROWID` without measuring any performance improvements, that's the extent of me considering it.